Send received Android onNewIntent to Linking.js

rreusser opened the pull request:
**Motivation:** This PR references the deep linking issue in #7079. `IntentAndroid` states that it is deprecated in favor of Linking, but Linking states that `addEventListener` and `removeEventListener` are [not supported in Android](https://github.com/facebook/react-native/blob/master/Libraries/Linking/Linking.js#L122). This PR follows the template of `onBackPressed` and adds both methods by implementing the `onNewIntent` method for ReactActivity and passing the result through DeviceEventEmitter to Linking.js.

**Test plan** (needs feedback before considered complete):

You may find a repo that uses this PR [here](https://github.com/rreusser/react-native-android-intent-issue/tree/on-new-intent-solution). An intent is sent through the terminal as:

```bash
$ adb shell am start -a android.intent.action.VIEW -d 'androidintentissue://testtest' com.androidintentissue
```

Linking adds an event listener using (updated to reflect requested changes):

```javascript
Linking.addEventListener('url', (url) => { ... });
```

The intent is received as:

```js
url = "androidintentissue://testtest"
```

![linking-android](https://cloud.githubusercontent.com/assets/572717/14717508/e0e344b6-07bf-11e6-8407-6757e8b8f1f4.gif)

**Documentation changes**:

Added note on necessity of `launchMode="singleTask"` to Linking under android instructions.

**Questions**:

1. Is this functionality desired? It seems useful but straightforward enough that perhaps it was left out for a reason.
2. Intents are not as simple as deep links, so I'm currently passing back the metadata as an object. This does not match the iOS API that passes back a simple `uri` string. I'll think about this a bit more, but improvements to this format are welcome.
3. I've documented that for Linking in android, the `launchMode` must be set to `"singleTask"` in order to prevent opening up multiple instances of `MainActivity`. Are there other considerations with the UI that affect this?
4. Other thoughts?

Unsolicited PR, so feel free to accept or reject, but we struggled with this a bit, so if there's anything I can do to help with this feature, I'd be glad to know more details/opinions/background. Thanks!

charpeni commented:
Sounds interesting !

satya164 commented:
@rreusser 

> Is this functionality desired? It seems useful but straightforward enough that perhaps it was left out for a reason.

No reason really.

> Intents are not as simple as deep links, so I'm currently passing back the metadata as an object. This does not match the iOS API that passes back a simple uri string. I'll think about this a bit more, but improvements to this format are welcome.

`getInitialURL` resolves to a string. And I think the `url` event should also follow the same format. I don't think we really need it to be an object.

1. action is always going to be `Intent.action.VIEW`. If not, it's not really a request to open a link. So you can drop the `action` prop.
2. `scheme` is already a part of the URI, so no need to pass it as a separate prop.
3. `type` is always undefined unless you're sharing content to the app, in which case, it shouldn't be handled by `Linking`.
4. `package` is also not needed

I think it can be just a string and cover majority of use cases, while being consistent with iOS. It's very confusing to have different results depending on the platform.

> I've documented that for Linking in android, the launchMode must be set to "singleTask" in order to prevent opening up multiple instances of MainActivity. Are there other considerations with the UI that affect this?

It entirely depends on whether individual needs. But we should document this.

dimitrovskif commented:
The feature definitely makes sense, and the `singleTask` mode should be documented but not set as default.

rreusser commented:
Thanks for the feedback! That makes sense. So ignore the intent silently in `onNewIntent` if it doesn't match `Intent.action.VIEW`?

satya164 commented:
@rreusser Yes, it should match the behaviour of `getInitialURL` - https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/modules/intent/IntentModule.java#L53

satya164 commented on ReactAndroid/src/main/java/com/facebook/react/ReactActivity.java:
```diff
@@ -223,4 +223,13 @@ public void onBackPressed() {
   }
+
+  @Override
+  public void onNewIntent(Intent intent) {
```
@andreicoman11 Any suggestions on how to handle this without needing to add the event to `ReactInstanceManager`?

mkonicek replied:
Can you add Javadoc explaining why this is needed? When someone sees this code later they won't see the description of this pull request anymore.

satya164 commented:
cc @mkonicek

rreusser commented:
Updated the PR to filter on `VIEW` and send uri as a string. Entering holding pattern on IntentModule feedback.

<img width="580" alt="screen shot 2016-04-21 at 2 23 51 pm" src="https://cloud.githubusercontent.com/assets/572717/14719909/d3ea3b68-07cc-11e6-973a-f6a4174a09d3.png">

rreusser commented:
I added some weasel words to the Linking docs to reflect singleTask may not always apply:

> NOTE: For instructions on how to add support for deep linking on Android, refer [Enabling Deep Links for App Content - Add Intent Filters for Your Deep Links](http://developer.android.com/training/app-indexing/deep-linking.html#adding-filters). If you wish to receive the intent in an existing instance of MainActivity, you may set the `launchMode` of MainActivity to `singleTask` in
`AndroidManifest.xml`. See [`<activity>`](http://developer.android.com/guide/topics/manifest/activity-element.html) documentation for more information.

>  ```
    <activity
      android:name=".MainActivity"
      android:launchMode="singleTask">
  ```

satya164 commented on ReactAndroid/src/main/java/com/facebook/react/ReactInstanceManagerImpl.java:
```diff
@@ -466,6 +466,22 @@ private void invokeDefaultOnBackPressed() {
+      String action = intent.getAction();
+
+      if (Intent.ACTION_VIEW.equals(action)) {
+        DeviceEventManagerModule deviceEventManagerModule =
```
We should check for if uri is null, and shouldn't send the event in that case. Can you copy the code from here instead? https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/modules/intent/IntentModule.java#L51

rreusser replied:
Modified. Pardon the sloppiness. Got in a hurry since need to check in on our Android release. If at any point it's easier to put this on a back burner or just make it right without trying to steer me in the right direction, don't hesitate 👍

grabbou commented:
Looks really cool! Exciting to see you managed to give it a go! Will try to find time tomorrow to check it out :)

satya164 commented:
@rreusser Awesome work. Let's wait for @mkonicek and @andreicoman11 before shipping.

rreusser commented:
@satya164 Thanks so much for the feedback! Y'all do an incredible job of keeping up with this. Yeah, IntentModule seems like a nice place for this to live, but it wasn't immediately clear how to route it through that, so glad to wait and see.

andreicoman11 commented on ReactAndroid/src/main/java/com/facebook/react/ReactInstanceManager.java:
```diff
@@ -88,6 +88,11 @@
   /**
+   * This method will give JS the opportunity to receive intents via Linking.
+   */
+  public abstract void onNewIntent(Intent intent);
```
Why does this live here instead of a module?

satya164 replied:
@andreicoman11 It doesn't exist in the `ActivityEventListener` interface, so can't be used in a module. Should we add it there?

andreicoman11 replied:
It looks like `onActivityResults` is also going through ReactInstanceManager, then I think this can follow the same route. Never mind, carry on.

satya164 replied:
@andreicoman11 Should we add this to https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/bridge/ActivityEventListener.java so that all modules can use it?

andreicoman11 replied:
I think this is good here. We should implement this for the new bridge as well (you can consider current bridge is deprecated). That would mean copying the code from `ReactInstanceManagerImpl` to https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/XReactInstanceManagerImpl.java.

mkonicek commented on ReactAndroid/src/main/java/com/facebook/react/modules/core/DeviceEventManagerModule.java:
```diff
@@ -53,6 +56,15 @@ public void emitHardwareBackPressed() {
   /**
+   * Sends an event to the JS instance that a new intent was received.
+   */
+  public void emitNewIntent(Uri uri) {
```
The docs are good!

Should this be called `emitNewIntentReceived` for clarity? By looking at the name itself I can't tell whether this creates a new intent somehow or whether it notifies JS the intent was received.

rreusser replied:
Agreed.

mkonicek commented on ReactAndroid/src/main/java/com/facebook/react/modules/core/DeviceEventManagerModule.java:
```diff
@@ -53,6 +56,15 @@ public void emitHardwareBackPressed() {
+  public void emitNewIntent(Uri uri) {
+    getReactApplicationContext()
+        .getJSModule(RCTDeviceEventEmitter.class)
+        .emit("openURL", uri.toString());
```
Why is this called "openURL"? If all of this logic works with generic URIs and Intents, this should be "intentReceived". Otherwise the Java method should be called `emitOpenURL` or similar.

satya164 replied:
@mkonicek The event is called `openURL` on iOS, so keeping same name means no extra handling for android is needed on JS side.

rreusser replied:
Yeah, as @satya164 said, it felt like there should be an intermediate layer (ActivityEventListener?) to which it sends the intent, and then that module makes the choice to interpret it as an opened URL, but that pathway wasn't quite clear.

mkonicek commented on ReactAndroid/src/main/java/com/facebook/react/ReactInstanceManagerImpl.java:
```diff
@@ -466,6 +467,23 @@ private void invokeDefaultOnBackPressed() {
+  @Override
+  public void onNewIntent(Intent intent) {
+    if (mCurrentReactContext == null) {
+      // Swallow the new intent?
```
nit: remove ? - if we commit this we've made that decision to swallow it :)

rreusser replied:
Yup. Will remove as soon as I have Internet!

mkonicek commented on Libraries/Linking/Linking.js:
```diff
@@ -46,9 +46,19 @@ const DEVICE_NOTIF_EVENT = 'openURL';
- * refer [Enabling Deep Links for App Content - Add Intent Filters for Your Deep Links](http://developer.android.com/training/app-indexing/deep-linking.html#adding-filters).
- *
+NOTE: For instructions on how to add support for deep linking on Android,
+refer [Enabling Deep Links for App Content - Add Intent Filters for Your Deep Links](http://developer.android.com/training/app-indexing/deep-linking.html#adding-filters).
```
nit: refer -> refer to

mkonicek commented:
Thanks for working on this!

I have a very basic question (sorry!) - what's the difference between this and `getInitialURL`?

satya164 commented:
@mkonicek `getInitialURL` is supposed to return the URL which started the activity. If the activity later received any intents, we need an event to handle it. In the default set up, a new activity is started every time the app receives a new intent, which starts a new instance of `ReactInstanceManager`, so this event isn't really needed there. But when the `launchMode` is set to `singleTask`, the same activity can receive new intents instead of starting new activities. In this case, we need this event to handle incoming links.

facebook-github-bot-0 commented:
@satya164 would you mind taking a look at this pull request? It's been a while since the last commit was reviewed.

mkonicek commented:
Thanks for the explanation @satya164!

@rreusser Can you address the requested changes please?

facebook-github-bot-9 commented:
@rreusser updated the pull request.

rreusser commented:
Sorry for the delay. Just started a new job so kinda scattered. Renamed emitNewIntent -> emitNewIntentReceived, removed a comment, and fixed the doc typo.

facebook-github-bot-9 commented:
@rreusser do you have any updates for this pull request? It's been a while so we wanted to check in and see if you've looked at the requested changes.

mkonicek commented on Libraries/Linking/Linking.js:
```diff
@@ -139,20 +145,16 @@ class Linking {
    * @platform ios
    */
   static removeEventListener(type: string, handler: Function ) {
```
nit: remove space before `)` please

rreusser replied:
Fixed, but GH is in some sort of weird read-only mode… waiting to see how this pans out…

mkonicek commented:
Thanks for updating the pull request @rreusser. This looks good to me now, @satya164 what do you think?

rreusser commented:
I read through the docs and made the following additional fixes to reflect that add/remove will no longer be iOS-only. Let me know if this is incorrect in any way and I can refine.

In Linking.js:

`removeEventListener`: removed "@platform ios" (ref: [Linking.js:92](https://github.com/facebook/react-native/blob/master/Libraries/Linking/Linking.js#L92))
`addEventListener`: removed "@platform ios" (ref: [Linking.js:118](https://github.com/facebook/react-native/blob/master/Libraries/Linking/Linking.js#L118))
And just above in the instructions, removed "Note that this is only supported on iOS." (ref: [Linking.js:139](https://github.com/facebook/react-native/blob/master/Libraries/Linking/Linking.js#L139))

facebook-github-bot-0 commented:
@rreusser updated the pull request.

mkonicek commented:
There's a new potentially similar PR: https://github.com/facebook/react-native/pull/7739

satya164 commented:
LGTM. @rreuser Sorry, this took long. Can you rebase against master before we merge?

rreusser commented:
Sure. Out for Memorial Day but will do, hopefully tomorrow.

facebook-github-bot-6 commented:
@rreusser updated the pull request.

rreusser commented:
@satya164 @mkonicek Rebased against `master`. Mainly just had to change event name `"openURL"` to `"url"`. Ran into android build issues\* so have *not* tested the merged version end to end. Scrambling a bit to avoid falling behind on the new job, so that's a bit more time than I can spare at the moment.

\* Testing against sample repo [here](https://github.com/rreusser/react-native-android-intent-issue/tree/on-new-intent-solution). Got: `Android NDK: Android.mk: Cannot find module with tag 'cxxreact' in import path; Android NDK: Are you sure your NDK_MODULE_PATH variable is properly defined ?` 😞

skyride99 commented:
Is this going to be in 0.27 of RN

satya164 commented:
@rreusser Sorry, missed your comment. Do you have the correct version of NDK configured? It's `r10e` I think.

satya164 commented:
Seems there are still conflicts with master :(

satya164 commented:
I think `openURL` is a better name for the event rather than `url`.

satya164 commented:
@skyride99 No. Most likely the next release if it gets merged before the branch cut.

facebook-github-bot-9 commented:
@rreusser updated the pull request.

facebook-github-bot-2 commented:
@rreusser updated the pull request.

rreusser commented:
@satya164 Fixed the merge conflict. Really wanted to confirm with a build, but hitting a brick wall on this one:

```
:ReactAndroid:buildReactNdkLib
Android NDK: WARNING: Ignoring unknown import directory: /Users/rreusser/projects/react-native/AndroidIntentIssue/AndroidIntentIssue/node_modules/react-native/ReactAndroid/../ReactCommon    
Android.mk:35: *** Android NDK: Aborting.    .  Stop.
Android NDK: Android.mk: Cannot find module with tag 'cxxreact' in import path    
Android NDK: Are you sure your NDK_MODULE_PATH variable is properly defined ?    
Android NDK: The following directories were searched:    
Android NDK:         
make: Entering directory `/Users/rreusser/projects/react-native/AndroidIntentIssue/AndroidIntentIssue/node_modules/react-native/ReactAndroid/src/main/jni/xreact/jni'
make: Leaving directory `/Users/rreusser/projects/react-native/AndroidIntentIssue/AndroidIntentIssue/node_modules/react-native/ReactAndroid/src/main/jni/xreact/jni'
:ReactAndroid:buildReactNdkLib FAILED
```

Rebuilt the android project with `react-native init` (0.26) and tried some variations of setting `NDK_MODULE_PATH`, but just hitting a brick wall on this since I don't understand the build process better. It's probably a one-liner `export`; I just don't happen to know what that one line is. 😞

satya164 commented:
@rreusser To confirm, what's the version of Android NDK you're using?

rreusser commented:
@satya164 `r10e`

satya164 commented:
@rreusser https://github.com/facebook/react-native/issues/7911

Do you have the NDK path set in `local.properties` also? I had to do that.

satya164 commented:
@facebook-github-bot shipit

facebook-github-bot-5 commented:
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/1012113985532642/int_phab) to review.

satya164 commented:
@rreusser Can you send another PR to fix `getInitialURL` so that it only returns the initial URL instead of the last received URL?

satya164 commented:
@facebook-github-bot shipit

facebook-github-bot-3 commented:
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/1012113985532642/int_phab) to review.

satya164 commented:
Seems the tests are failing. The code doesn't compile. Will check when I get some time.

satya164 commented:
I fixed the compilation issue and opened another PR #7940

skyride99 commented:
Will it make it 0.28?

satya164 commented:
@skyride99 I've shipped the PR. Let's hope the merge is done before @grabbou does the branch cut :D

satya164 commented:
Closing this in favor of #7940

andreicoman11 commented on ReactAndroid/src/main/java/com/facebook/react/ReactInstanceManagerImpl.java:
```diff
@@ -432,6 +433,22 @@ private void invokeDefaultOnBackPressed() {
+  @Override
+  public void onNewIntent(Intent intent) {
+    if (mCurrentReactContext == null) {
+      FLog.w(ReactConstants.TAG, "Instance detached from instance manager");
```
nit: `return;` after Flog.w in order to simplify the `else` branch.
